Detectors/TRD/base/src/LTUParam.cxx

```diff
@@ -0,0 +1,249 @@
+  // i.e. Lorentz angle and tilt correction (if active)
+
+  int layer = det % 6;
+
+  float dyTilt = (mgDriftLength * std::tan(mgTiltingAngle[layer] * M_PI / 180.) *
+                  getLocalZ(det, rob, mcm) * mgInvX[layer]);
+
+  // calculate Lorentz correction
+  float dyCorr = -mOmegaTau * mgDriftLength;
+
+  if (mTiltCorr)
+    dyCorr += dyTilt; // add tilt correction
+
+  return (int)TMath::Nint(dyCorr * mgScalePad * mgInvWidthPad[layer]);
+}
+
+void LTUParam::getDyRange(int det, int rob, int mcm, int ch,
+                          int& dyMinInt, int& dyMaxInt) const
+{
+  // calculate the deflection range in which tracklets are accepted
+
+  dyMinInt = mgDyMin;
+  dyMaxInt = mgDyMax;
+
+  // deflection cut is considered for |B| > 0.1 T only
+  if (TMath::Abs(mMagField) < 0.1)
+    return;
+
+  float e = 0.30;
+
+  float maxDeflTemp = getPerp(det, rob, mcm, ch) / 2. *               // Sekante/2 (cm)
+                      (e * 1e-2 * TMath::Abs(mMagField) * mInvPtMin); // 1/R (1/cm)
+
+  float phi = getPhi(det, rob, mcm, ch);
+  if (maxDeflTemp < TMath::Cos(phi)) {
+    float maxDeflAngle = TMath::ASin(maxDeflTemp);
+
+    float dyMin = (mgDriftLength *
+                   std::tan(phi - maxDeflAngle));
+
+    dyMinInt = int(dyMin / mgBinDy);
+    // clipping to allowed range
+    if (dyMinInt < mgDyMin)
+      dyMinInt = mgDyMin;
+    else if (dyMinInt > mgDyMax)
+      dyMinInt = mgDyMax;
+
+    float dyMax = (mgDriftLength *
+                   std::tan(phi + maxDeflAngle));
+
+    dyMaxInt = int(dyMax / mgBinDy);
+    // clipping to allowed range
+    if (dyMaxInt > mgDyMax)
+      dyMaxInt = mgDyMax;
+    else if (dyMaxInt < mgDyMin)
+      dyMaxInt = mgDyMin;
+  } else if (maxDeflTemp < 0.) {
+    // this must not happen
+    printf("Inconsistent calculation of sin(alpha): %f\n", maxDeflTemp);
+  } else {
+    // TRD is not reached at the given pt threshold
+    // max range
+  }
+
+  if ((dyMaxInt - dyMinInt) <= 0) {
+    LOG(info) << "strange dy range: [" << dyMinInt << "," << dyMaxInt << "], using max range now";
+    dyMaxInt = mgDyMax;
+    dyMinInt = mgDyMin;
+  }
+}
+
+float LTUParam::getElongation(int det, int rob, int mcm, int ch) const
+{
+  // calculate the ratio of the distance to the primary vertex and the
+  // distance in x-direction for the given ADC channel
+
+  int layer = det % 6;
+
+  float elongation = TMath::Abs(getDist(det, rob, mcm, ch) * mgInvX[layer]);
+
+  // sanity check
+  if (elongation < 0.001) {
+    elongation = 1.;
+  }
+  return elongation;
+}
+
+void LTUParam::getCorrectionFactors(int det, int rob, int mcm, int ch,
+                                    unsigned int& cor0, unsigned int& cor1, float gain) const
+{
+  // calculate the gain correction factors for the given ADC channel
+  float Invgain = 1.0;
+  if (mPidGainCorr == true)
+    Invgain = 1 / gain;
+
+  if (mPidTracklengthCorr == true) {
+    cor0 = (unsigned int)((1.0 * mScaleQ0 * (1. / getElongation(det, rob, mcm, ch))) * Invgain);
```

Comment:
getElongation could be cached (only called once)

Reply:
and the 1.0 * and the division and the end gain multiplication. 
In looking for other redundant calculations I saw I was doing an X^4 not and X^2 at line 231 in getPerp.
I also reverted back to the for_each's and lambdas in the constructor.